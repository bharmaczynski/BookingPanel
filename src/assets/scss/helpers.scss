$font-size-s: 14px;
$font-size-base: 16px;
$font-size-l: 20px;
$font-size-xl: 24px;
$font-size-xxl: 30px;

.font-size-s {
  font-size: $font-size-s;
}

.font-size-base {
  font-size: $font-size-base;
}

.font-size-l {
  font-size: $font-size-l;
}

.font-size-xl {
  font-size: $font-size-xl;
}

.font-size-xxl {
  font-size: $font-size-xxl;
}

.text {
  &--bold {
    font-weight: bold;
  }
}
