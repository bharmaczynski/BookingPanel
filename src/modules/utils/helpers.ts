import moment from 'moment';

export const isBetweenDates = (
    dateStart: string,
    dateEnd: string,
    dateToCheck: string
): boolean => {
    // included dateEnd
    return (
        moment(dateToCheck, 'DD-MM-YYYY').diff(moment(dateStart, 'DD-MM-YYYY'), 'days') > 0 &&
        moment(dateToCheck, 'DD-MM-YYYY').diff(moment(dateEnd, 'DD-MM-YYYY'), 'days') <= 0
    );
};

export const isPastDay = (date: string): boolean => {
    return moment(date, 'DD-MM-YYYY').diff(moment(), 'days') < 0;
};

export const isToday = (date: string): boolean => {
    return date === moment().format('DD-MM-YYYY');
};
