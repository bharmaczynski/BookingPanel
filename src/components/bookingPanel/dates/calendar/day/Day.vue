<template>
    <div
        class="day"
        :class="{
            'day--inactive': day.outOfMonth,
            'day--today': day.isToday,
            'day--unavailable':
                setPastDay(day.value) || day.unavailable || day.temporaryUnavailable,
            'day--start': day.value === checkedDate.checkIn,
            'day--end': day.value === checkedDate.checkOut,
            'day--between': day.isBetween,
            'day--invalid': day.isInvalid,
        }"
        @click="handleDayClick(day.value)"
        @mouseenter="handleMouseEnter(day.value)"
    >
        <span class="day__text font-size-s">{{ day.number }}</span>
    </div>
</template>
<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import IDay from '@/ts/types/day.type';
import { isPastDay } from '@/modules/utils/helpers';
import IDate from '@/ts/types/date.type.ts';

@Component
export default class Day extends Vue {
    @Prop({ required: true })
    day!: IDay;

    @Prop({ default: { checkIn: '', checkOut: '' } })
    checkedDate!: IDate;

    handleDayClick(value: string): void {
        this.$emit('click', value);
    }

    handleMouseEnter(value: string): void {
        this.$emit('mouseEnter', value);
    }

    setPastDay(date: string): boolean {
        return isPastDay(date);
    }
}
</script>
<style lang="scss" scoped>
.day {
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 38px;
    height: 38px;
    border-radius: 50%;
    cursor: pointer;
    border: 2px solid transparent;

    &:hover {
        background: #cffcf8;
        color: $downy;

        &::before {
            right: 19px !important;
        }
    }

    &--inactive {
        color: #888;
    }

    &--today {
        color: $downy;
        border: 2px solid $downy;
    }

    &--unavailable {
        color: rgba(#333, 0.3);
        pointer-events: none;
    }

    &--between {
        color: $downy;
        background: #cffcf8;

        &::before {
            content: '';
            display: block;
            position: absolute;
            z-index: -1;
            top: -2px;
            right: 19px;
            bottom: -2px;
            left: -19px;
            background: #cffcf8;
        }
    }

    &--start,
    &--end {
        background: $downy;
        color: #fff;
    }

    &--invalid {
        color: #ca0b00;
        border: 2px solid red;
    }

    &__text {
        font-weight: 500;
    }
}
</style>
