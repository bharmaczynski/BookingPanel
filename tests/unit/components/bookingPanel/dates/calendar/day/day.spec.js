import { shallowMount } from '@vue/test-utils';
import Day from '@/components/bookingPanel/dates/calendar/day/Day.vue';
import * as helpers from '@/modules/utils/helpers';

describe('Day', () => {
    let propsData;

    beforeEach(() => {
        propsData = {
            day: {
                number: 2,
                outOfMonth: false,
                isToday: false,
                unavailable: false,
                value: '02-02-2021',
                isBetween: false,
                temporaryUnavailable: false,
                isInvalid: false,
            },
            checkedDate: {
                checkIn: '10-10-2021',
                checkOut: '12-10-2021',
            },
        };
    });

    function createWrapper(propsData = {}) {
        return shallowMount(Day, {
            propsData,
        });
    }

    describe('Snapshots', () => {
        it('should match initial snapshot', () => {
            const wrapper = createWrapper(propsData);

            expect(wrapper.vm.$el).toMatchSnapshot();
        });
    });

    describe('Methods', () => {
        let wrapper;
        let isPastDayMock = true;

        beforeEach(() => {
            wrapper = createWrapper(propsData);
            wrapper.vm.$emit = jest.fn();
            helpers.isPastDay = jest.fn(() => isPastDayMock);
        });

        describe('handleDayClick', () => {
            it('should emit event with proper value', () => {
                wrapper.vm.handleDayClick('02-02-2021');

                expect(wrapper.vm.$emit).toHaveBeenCalledWith('click', '02-02-2021');
            });
        });

        describe('handleMouseEnter', () => {
            it('should emit event with proper value', () => {
                wrapper.vm.handleMouseEnter('02-02-2021');

                expect(wrapper.vm.$emit).toHaveBeenCalledWith('mouseEnter', '02-02-2021');
            });
        });

        describe('setPastDay', () => {
            it('should return true in case helpers.isPastDay is true', () => {
                expect(wrapper.vm.setPastDay('12-12-2002')).toBe(true);
            });

            it('should return proper value', () => {
                isPastDayMock = false;
                expect(wrapper.vm.setPastDay('12-12-2002')).toBe(false);
            });
        });
    });
});
