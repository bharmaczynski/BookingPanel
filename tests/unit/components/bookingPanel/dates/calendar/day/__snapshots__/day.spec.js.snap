// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day Snapshots should match initial snapshot 1`] = `
<div
  class="day day--unavailable"
>
  <span
    class="day__text font-size-s"
  >
    2
  </span>
</div>
`;
